src-tauri/src/main.rs

```diff
@@ -100,13 +119,93 @@ fn main() {
       // TODO This should hide the main taskbar icon when the window is closed on macos, however there is a borrow error
       // #[cfg(target_os = "macos")]
       // app.set_activation_policy(tauri::ActivationPolicy::Accessory);
-      let tray_handle = app_handle.tray_handle().clone();
-      let item_handle = tray_handle.get_item(&"toggle_visibility");
+      let tray_handle = app_handle.tray_handle();
+      let item_handle = tray_handle.get_item("toggle_visibility");
       item_handle.set_title("Show").unwrap(); // update the tray menu title to reflect the hidden state of the window
     }
     Event::ExitRequested { api, .. } => {
       api.prevent_exit();
     }
     _ => {}
+  });
+
+  Ok(())
+}
+
+/// Start farming by using plot in specified path and connecting to WebSocket server at specified
+/// address.
+pub(crate) async fn farm(
+  base_directory: PathBuf,
+  node_rpc_url: &str,
+) -> Result<[u8; 32], anyhow::Error> {
+  // TODO: This doesn't account for the fact that node can
+  // have a completely different history to what farmer expects
+  info!("Opening plot");
+  let plot_fut = tokio::task::spawn_blocking({
+    let base_directory = base_directory.clone();
+
+    move || Plot::open_or_create(&base_directory)
+  });
+  let plot = plot_fut.await.unwrap()?;
+
+  info!("Opening commitments");
+  let commitments_fut = tokio::task::spawn_blocking({
+    let path = base_directory.join("commitments");
+
+    move || Commitments::new(path)
+  });
+  let commitments = commitments_fut.await.unwrap()?;
+
+  info!("Opening object mapping");
+  let object_mappings = tokio::task::spawn_blocking({
+    let base_directory = base_directory.clone();
+
+    move || ObjectMappings::open_or_create(&base_directory)
   })
+  .await??;
+
+  info!("Connecting to node at {}", node_rpc_url);
+  let client = WsRpc::new(node_rpc_url).await?;
+
+  let farmer_metadata = client
+    .farmer_metadata()
+    .await
+    .map_err(|error| anyhow::Error::msg(error.to_string()))?;
+
+  let identity = Identity::open_or_create(&base_directory)?;
+
+  let subspace_codec = SubspaceCodec::new(identity.public_key());
+
+  // start the farming task
+  let farming_instance = Farming::start(
+    plot.clone(),
+    commitments.clone(),
+    client.clone(),
+    identity.clone(),
+  );
+
+  // start the background plotting
+  let plotting_instance = Plotting::start(
+    plot,
+    commitments,
+    object_mappings,
+    client,
+    farmer_metadata,
+    subspace_codec,
+  );
+
+  // wait for the farming and plotting in the background
+  tokio::spawn(async {
+    tokio::select! {
+      res = plotting_instance.wait() => if let Err(error) = res {
+        return Err(anyhow!(error))
+      },
+      res = farming_instance.wait() => if let Err(error) = res {
+        return Err(anyhow!(error))
+      },
```

Comment:
But these are not even printed anywhere, right?

Reply:
currently no. They are like a placeholder at the moment, to remind my future self, it is possible to do something with them, or maybe get rid of them :)

Reply:
TODO would make it clean :wink: